Std/Data/MLList/Heartbeats.lean

```diff
@@ -0,0 +1,24 @@
+/-
+Copyright (c) 2023 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import Std.Data.MLList.Basic
+import Std.Lean.CoreM
+
+/-!
+# Truncate a `MLList` when running out of available heartbeats.
+-/
+
+open Lean.Core (CoreM)
+
+/-- Take an initial segment of a monadic lazy list,
+trying to leave at least `percent` of the remaining allowed heartbeats. -/
```

Comment:
This description is not so good. When is the heartbeat check run, and what will be in the list? The doc should give a precise specification without terms like "trying to".

Reply:
Excellent criticism! The code from Mathlib was actually discarding the last produced element if it went over the time limit, for no good reason.

I've changed it to use `takeUpToFirstM` rather than `takeWhileM`, and expanded the doc-string.